### DISCUSSION TITLE
How to unbind @Singleton

### DISCUSSION THREAD

**Discussion by TheDeadOne** [playframework/playframework#12936] [Q&A] [answered]
I have a monolithic application with a huge ammount of classes annotated with `@Singleton`. Their automatic loading, without the need to write walls of `bind().to()` in modules, is convenient until the need for integration tests arises. In a test, I don't need all the components of the application, only dependencies of this particular test. How to prevent binding of annotated classes into the context or unbind unnecessary ones?

**Comment by TheDeadOne** [comment]
I found part of the answer, [implicit binding needs to be disabled](https://github.com/google/guice/wiki/JustInTimeBindings#enforce-explicit-bindings). But now I get errors like

```
Explicit bindings are required and controllers.Application is not explicitly bound.
  while locating controllers.Application
    for the 2nd parameter of router.Routes.<init>(Routes.scala:312)
  at play.api.inject.RoutesProvider$.bindingsFromConfiguration(BuiltinModule.scala:129):
Binding(class router.Routes to self) (via modules: com.google.inject.util.Modules$OverrideModule ->
  play.api.inject.guice.GuiceableModuleConversions$$anon$4)
```

I can't replace the routes using the `play.http.router` parameter because the code has too much reverse routing and it won't compile. Is it possible to compile the router but not load it into the Guice context?

**Comment by TheDeadOne** [comment, marked as answer]
I copied source of the [BuiltinModule](https://github.com/playframework/playframework/blob/2.7.x/core/play/src/main/scala/play/api/inject/BuiltinModule.scala) module, replace the call to the `RoutesProvider.bindingsFromConfiguration` method with `FakeRouterProvider.bindingStub`, which returns an empty router, disabled the original module in the configuration and enabled patched one. I don’t know if this is right or if it can’t be done better, but it works.

**Reply by ihostage** [reply]
Hi, Sergey! 
Did you try this [override approach](https://www.playframework.com/documentation/3.0.x/JavaTestingWithGuice#Override-bindings)?

**Reply by TheDeadOne** [reply]
Yes, but without success. BuiltinModule creates an instance of GeneratedRouted [using reflection](https://github.com/playframework/playframework/blob/2.7.x/core/play/src/main/scala/play/api/routing/Router.scala#L88) and then [adds the instance](https://github.com/playframework/playframework/blob/2.7.x/core/play/src/main/scala/play/api/inject/BuiltinModule.scala#L129) to the Guice context, and GeneratedRouter calls the constructors of all controllers that have routes. So it cannot be overridden in GuiceApplicationBuilder, as I see.
